### ISSUE TITLE
DNS queries from router are not forwarded to container's private interface

### ISSUE THREAD

**Issue by mfschumann** [containers/podman#8195]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

I am trying to run a pihole DNS server in a rootful container with --net=private on my Fedora server (release 33). My internet gateway router is configured to use this DNS server as upstream, so that clients on my network would usually send their DNS queries to the router who forwards them to the DNS server running in the container.

The issue is that clients get replies to their DNS requests only if they are contacting the DNS server directly, but never when contacting it via the router.

I've tracked the problem down by tcpdumping the ethernet interface of the Fedora server and the container's private interface while sending DNS queries. I found that both types of DNS queries (direct and via the router) appear on the ethernet interface, but only the direct DNS queries make it to the container's private interface. Some misconfiguration/feature/bug of podman seems to block the router's DNS queries from being forwarded between the two interfaces.

**Steps to reproduce the issue:**

1. Run a rootful container that listens on port 53 of the host with --net=private.

2. Send DNS queries from any network client directly to the host.

3. Send DNS queries from a network client via the internet gateway router to the host.

**Describe the results you received:**
DNS queries sent via the internet gateway router are not forwarded to the container's private interface, while DNS queries sent directly do reach the private interface.

**Describe the results you expected:**
Both DNS queries are forwarded to the container's private interface.

**Additional information you deem important (e.g. issue happens only occasionally):**
I am using firewalld on the host, but I don't see why it should block traffic depending on source address.

**Output of `podman version`:**

```
Version:      2.1.1
API Version:  2.0.0
Go Version:   go1.15.2
Built:        Wed Oct  7 18:21:20 2020
OS/Arch:      linux/amd64
```

**Output of `podman info --debug`:**

```
host:
  arch: amd64
  buildahVersion: 1.16.1
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: conmon-2.0.21-3.fc33.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.21, commit: 0f53fb68333bdead5fe4dc5175703e22cf9882ab'
  cpus: 2
  distribution:
    distribution: fedora
    version: "33"
  eventLogger: journald
  hostname: srv
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 5.8.16-300.fc33.x86_64
  linkmode: dynamic
  memFree: 4267311104
  memTotal: 8328482816
  ociRuntime:
    name: crun
    package: crun-0.15-5.fc33.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 0.15
      commit: 56ca95e61639510c7dbd39ff512f80f626404969
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
  os: linux
  remoteSocket:
    exists: true
    path: /run/podman/podman.sock
  rootless: false
  slirp4netns:
    executable: ""
    package: ""
    version: ""
  swapFree: 4163891200
  swapTotal: 4163891200
  uptime: 24m 45.96s
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - registry.centos.org
  - docker.io
  srv.mfs.name:5000:
    Blocked: false
    Insecure: true
    Location: srv.mfs.name:5000
    MirrorByDigestOnly: false
    Mirrors: null
    Prefix: srv.mfs.name:5000
store:
  configFile: /etc/containers/storage.conf
  containerStore:
    number: 1
    paused: 0
    running: 0
    stopped: 1
  graphDriverName: overlay
  graphOptions:
    overlay.mountopt: nodev,metacopy=on
  graphRoot: /var/lib/containers/storage
  graphStatus:
    Backing Filesystem: btrfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "true"
  imageStore:
    number: 2
  runRoot: /var/run/containers/storage
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 2.0.0
  Built: 1602087680
  BuiltTime: Wed Oct  7 18:21:20 2020
  GitCommit: ""
  GoVersion: go1.15.2
  OsArch: linux/amd64
  Version: 2.1.1
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman-2.1.1-12.fc33.x86_64
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide?**
Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**
The host is running on bare metal.

**Comment by usury** [comment]
I have some familiarity with firewalld on Fedora 32, @mfschumann. I had to learn some new things regarding kernel api "netfilter" (new) vs "getsocket/setsocket" (old) and iptables-nft (new) vs iptables-legacy (old). I had been using "iptables" since the days of "FC3" when we had to recompile our own kernels in order to get certain flexability (like ip masquerade, etc).

**Depending on whether your Fedora is a fresh install or an upgrade over the years, it is _possible_ that something exists in one of "legacy" tables that blocks packets before they reach your rules configured in firewalld.**

By default, firewalld uses the "netfilter" backend, which is an important observation if you are accustomed to using "iptables" from the commandline to see your ruleset. It took me awhile to understand why I was unable to see my "firewalld" rules when using the familiar `root@fedora #> iptables -t filter -nvL`

You'll notice that /usr/sbin/ no longer has an iptables binary but rather various aliases for iptables, controlled by the "alternatives" mechanism
```
root@fedora #> ls -l /usr/sbin/iptables*
root@fedora #> ls -l /usr/sbin/ip6tables*
root@fedora #> ls -l $(readlink -f $(which iptables))
root@fedora #> dnf whatprovides $(readlink -f $(which iptables))
root@fedora #> dnf whatprovides /usr/sbin/xtables-nft-multi
```
By default, my fresh install of Fedora 32 had only the `xtables-legacy-multi` binary (wheras my Debian systems have both, installed by the same package even).

However, after installing the "iptables-nft" packge, I am still unable to use "iptable-nft" to see rules configured by firewalld. That's a mystery to me. The man page for "xtables-nft" pointed me to `root@fedora #> nft list ruleset`. Of course, I had to install the "nft" package first since it also was not installed by default.

After installing "iptables-nft", you'll probably want to run `root@fedora #> alternatives --config iptables` to select the one `root@fedora #> iptables` invokes.

Maybe you already knew all that. Maybe there are no "legacy" rules blocking inbound DNS requests originating from your router. It's worth confirming.

Regardless, you will love the pihole setup once it is working properly. I run it on a stand-alone raspberry pi. It's fantastic.

**Comment by mfschumann** [comment]
Tracing the activities of nft with `nft monitor trace` I found that packets coming from a client were designated `daddr 10.88.0.61` (the container's private IP), while packets coming from the router were designated `daddr 192.168.0.50` (the host's LAN IP). This difference was caused by a firewalld rule that opened port 53 on the host. Once I deleted the rule the packets from the router showed up with `daddr 10.88.0.61` and made their way into the container.

In conclusion: Opening a host port via firewalld while binding that port to a container prevents packets sent by the host's gateway from flowing into the container. Is this behavior expected / wanted?

**Comment by usury** [comment]
> In conclusion: Opening a host port via firewalld while binding that port to a container prevents packets sent by the host's gateway from flowing into the container. Is this behavior expected / wanted?

I'm a relatively new podman user, so I'm in no position to speak with authority.

If "podman" is adding some firewall rule (by calling "iptables-legacy", "iptables-nft", or some other more direct mechanism) that interfers with otherwise intended behavior, that sounds like something for podman to handle differently or at least document.

If it is the user adding a rule via "firewalld" that interfers with the default behavior of "podman" (regarding publishing/forwarding container ports), then that sounds like a problem outside the scope of "podman".

**Comment by rhatdan** [comment]
@mccv1r0 @mheon PTAL

**Comment by rhatdan** [comment]
Podman relies on CNI for configuration of the network stack.

**Comment by mfschumann** [comment]
Even with the firewalld port closed, I found that operation was not reliable: Many packets from the router still appeared in the nft trace with the host IP as destination, and when re-running the pihole container (so it got assigned a new private IP), a fraction of the packets was destined to the private IP the container had before the restart. All in all, DNS resolution via the router worked most of the time, but was often very slow.

I ended up replacing the host-local IPAM configuration by a static configuration, so the PiHole container is now assigned a fixed private IP. Also, I added a static route in my router's configuration that sets the Fedora server as a gateway for the private subnet. Like that, I was able to set the upstream DNS server for my router to the now-static private IP of the PiHole container. With this setup, DNS resolution via the router finally works smoothly.

**Comment by mccv1r0** [comment]
CNI just configures Linux network with parameters supplied by podman, which gets these from user (typically via command line arguments.)

> I am trying to run a pihole DNS server in a rootful container with --net=private on my Fedora server (release 33). My internet gateway router is configured to use this DNS server as upstream, so that clients on my network would usually send their DNS queries to the router who forwards them to the DNS server running in the container.
> 
> The issue is that clients get replies to their DNS requests only if they are contacting the DNS server directly, but never when contacting it via the router.
> 
> **Steps to reproduce the issue:**
> 
> 1. Run a rootful container that listens on port 53 of the host with --net=private.
> 2. Send DNS queries from any network client directly to the host.
> 3. Send DNS queries from a network client via the internet gateway router to the host.

`--net=private` could be anything.  Can you supply the configuration?

re 1:  every time you start this container, it's IPv4 address will change.  Unless you make this static you will have to update your router to take account of the new address.  It reads like you have figured this out already. 

re: 3: We need to know the topology involved.  What's between the `internet gateway router` and podman `--net=private`

What is the default policy for the `FORWARD` chain?

I'm *guessing* that the `FORWARD` chain is dropping any packets from outside to anything on the bridge(???) used by `--net=private` which is a feature.  By design you can connect out (and RELATED) traffic is allowed back.  But traffic from outside to a pod is dropped.    This assumes bridge you are using CNI plugin, your router is reachable via `eth0` of your host where podman runs and your container is on Linux bridge (e.g. `cni0`).   The packets from `eth0` are routed (`eth0`->`cni0`), hitting the `FORWARD` chain as a result.  Without more information I can only guess.

The output of `iptables -nvL` and `iptables -nvL -t nat` would be helpful; as would the IP address of the container created by podman at the time those commands were run.

> Many packets from the router still appeared in the nft trace with the host IP as destination

This might be your DNS server configuration?   Not enough info, but what the outside world sends is up to configuration outside of podman.

> and when re-running the pihole container (so it got assigned a new private IP), a fraction of the packets was destined to the private IP the container had before the restart. 

See above

**Comment by mfschumann** [comment]
> `--net=private` could be anything. Can you supply the configuration?

You can find the CNI config that was active when the issues occured [here](https://pastebin.com/Xx3Leg2U).

> re 1: every time you start this container, it's IPv4 address will change. Unless you make this static you will have to update your router to take account of the new address. It reads like you have figured this out already.

Yes, with a static IP for the container things do work now. That means from my side you do not have to chase this issue further. In case you want to make sure it is not a bug that I was hitting I will provide the information you requested below.

> re: 3: We need to know the topology involved. What's between the `internet gateway router` and podman `--net=private`

The host is connected to the router via its enp0s25 interface with IP 192.168.0.50/24. The container was on a bridge with a host-local IP from the range 10.88.0.0/16.

> What is the default policy for the `FORWARD` chain?
>
> I'm _guessing_ that the `FORWARD` chain is dropping any packets from outside to anything on the bridge(???) used by `--net=private` which is a feature. By design you can connect out (and RELATED) traffic is allowed back. But traffic from outside to a pod is dropped. This assumes bridge you are using CNI plugin, your router is reachable via `eth0` of your host where podman runs and your container is on Linux bridge (e.g. `cni0`). The packets from `eth0` are routed (`eth0`->`cni0`), hitting the `FORWARD` chain as a result. Without more information I can only guess.

iptables does not have a `FORWARD` chain, but nft has one. And if I read the [output of](https://pastebin.com/7EKr6zcu) `nft list ruleset` correctly, it only forwards new packets with destination 10.88.0.91, but doesn't when they arrive with destination 192.168.0.50. So your guess was a hit it seems.

> The output of `iptables -nvL` and `iptables -nvL -t nat` would be helpful; as would the IP address of the container created by podman at the time those commands were run.

`iptables -nvL` does not show any rules. The output of `iptables -nvL -t nat` can be found [here](https://pastebin.com/Yd3jjUji).

In case you need further information, I am happy to provide it.

**Comment by rhatdan** [comment]
@baude PTAL

**Comment by baude** [comment]
i need @mccv1r0 on this, it is well beyond simple networking ....

**Comment by rhatdan** [comment]
@mccv1r0 Any thoughts on this one?

**Comment by rhatdan** [comment]
@mccv1r0 Reminder?

**Comment by baude** [comment]
I believe this issue should be closed with Podman-4.0.2 and the new network stack.  If you try POdman 4.0.2 from an upgrade to a previous version of podman, please be sure to run podman system reset --force before trying your test again.  Please re-open if new Podman does not resolve your issue.
